In [1]:
from dask import bag as db

In [2]:
import os
import operator as op
import re
import string
from collections import Counter
from itertools import chain

In [3]:
rus = set([chr(i) for i in range(ord("а"), ord("я") + 1)] + ['ё'])

In [4]:
files = [os.path.join('C:\\Users\\Big data\\New\\all_k', i) for i in os.listdir('C:\\Users\\Big data\\New\\all_k')]

In [5]:
def count_1(file_name):
    with open(file_name, encoding='ANSI') as f:
        cnt = Counter(chain.from_iterable((i for i in line.lower() if i in rus) for line in f))
    return cnt

In [6]:
first = db.from_sequence(files).map(count_1)

In [7]:
%%time
first.fold(operator.add).compute()

Wall time: 16.9 s


Counter({'х': 2041934,
         'у': 6891652,
         'л': 12464299,
         'и': 15863838,
         'о': 27133057,
         'к': 8034359,
         'р': 10644472,
         'т': 14783382,
         'а': 19640388,
         'с': 12951076,
         'м': 7763665,
         'д': 7547962,
         'е': 20448261,
         'ь': 4836178,
         'я': 5071866,
         'б': 4234156,
         'н': 16379210,
         'г': 4253717,
         'ч': 3592169,
         'в': 10199348,
         'з': 4107524,
         'э': 1055764,
         'п': 6726782,
         'ю': 1409189,
         'ы': 4514797,
         'ш': 1988269,
         'ц': 821233,
         'ж': 2637251,
         'й': 2695494,
         'ф': 424631,
         'щ': 838272,
         'ъ': 57419,
         'ё': 8203})

In [8]:
def count_2(line):
    cnt = Counter((i for i in line.lower() if i in rus))
    return cnt

In [9]:
files = r'C:\\Users\\Big data\\New\\all_k\\*.TXT'
second = db.read_text(files, encoding='ANSI')

In [10]:
second.take(10)

('         Хулио Кортасар\n',
 '       \n',
 '         62. Модель для сборки\n',
 '       \n',
 '          \n',
 '       \n',
 '         Многие читатели, наверно, заметят, что в этом произведении я  кое-где\n',
 '      преступаю литературные условности. Приведу лишь несколько примеров:  ар-\n',
 '      гентинцы у меня то и дело переходят с "вы" на "ты", когда  это  для  них\n',
 '      естественно в диалоге; житель  Лондона,  только  недавно  бравший  уроки\n')

In [11]:
%%time
second.map(count_2).fold(operator.add).compute()

Wall time: 31.8 s


Counter({'х': 2041934,
         'у': 6891652,
         'л': 12464299,
         'и': 15863838,
         'о': 27133057,
         'к': 8034359,
         'р': 10644472,
         'т': 14783382,
         'а': 19640388,
         'с': 12951076,
         'м': 7763665,
         'д': 7547962,
         'е': 20448261,
         'ь': 4836178,
         'я': 5071866,
         'б': 4234156,
         'н': 16379210,
         'г': 4253717,
         'ч': 3592169,
         'в': 10199348,
         'з': 4107524,
         'э': 1055764,
         'п': 6726782,
         'ю': 1409189,
         'ы': 4514797,
         'ш': 1988269,
         'ц': 821233,
         'ж': 2637251,
         'й': 2695494,
         'ф': 424631,
         'щ': 838272,
         'ъ': 57419,
         'ё': 8203})

In [31]:
def count_3(text):
    txts=re.findall('[а-яА-ЯёЁ]', text.lower())
    return Counter(txts)

In [32]:
%%time
data_path=r'C:\\Users\\Big data\\New\\all_k'
b=db.read_text(os.path.join(data_path, '*.TXT'), encoding='ansi')
print(b.map(count_3).fold(op.add).compute())

Counter({'о': 27133057, 'е': 20448261, 'а': 19640388, 'н': 16379210, 'и': 15863838, 'т': 14783382, 'с': 12951076, 'л': 12464299, 'р': 10644472, 'в': 10199348, 'к': 8034359, 'м': 7763665, 'д': 7547962, 'у': 6891652, 'п': 6726782, 'я': 5071866, 'ь': 4836178, 'ы': 4514797, 'г': 4253717, 'б': 4234156, 'з': 4107524, 'ч': 3592169, 'й': 2695494, 'ж': 2637251, 'х': 2041934, 'ш': 1988269, 'ю': 1409189, 'э': 1055764, 'щ': 838272, 'ц': 821233, 'ф': 424631, 'ъ': 57419})
Wall time: 36.3 s


In [35]:
%%time
b=db.read_text(os.path.join(data_path, '*.TXT'), encoding='ansi')
print(b.map(count_3).flatten().frequencies().compute())

[('х', 930864), ('у', 1629632), ('л', 1881058), ('и', 1885354), ('о', 1984782), ('к', 1662130), ('р', 1805506), ('т', 1885094), ('а', 1962818), ('с', 1845623), ('м', 1660938), ('д', 1679367), ('е', 1943754), ('ь', 1467349), ('я', 1490332), ('б', 1382214), ('н', 1914711), ('г', 1388755), ('ч', 1322094), ('в', 1771597), ('з', 1403340), ('э', 701680), ('п', 1639243), ('ю', 756541), ('ы', 1374847), ('ш', 994998), ('ц', 528186), ('ж', 1147736), ('й', 1081859), ('ф', 304703), ('щ', 538019), ('ъ', 54150)]
Wall time: 25 s


In [38]:
%%time
second.map(lambda x: re.findall('[а-я]', x.lower())).flatten().frequencies().compute()

Wall time: 24.2 s


[('х', 2041934),
 ('у', 6891652),
 ('л', 12464299),
 ('и', 15863838),
 ('о', 27133057),
 ('к', 8034359),
 ('р', 10644472),
 ('т', 14783382),
 ('а', 19640388),
 ('с', 12951076),
 ('м', 7763665),
 ('д', 7547962),
 ('е', 20448261),
 ('ь', 4836178),
 ('я', 5071866),
 ('б', 4234156),
 ('н', 16379210),
 ('г', 4253717),
 ('ч', 3592169),
 ('в', 10199348),
 ('з', 4107524),
 ('э', 1055764),
 ('п', 6726782),
 ('ю', 1409189),
 ('ы', 4514797),
 ('ш', 1988269),
 ('ц', 821233),
 ('ж', 2637251),
 ('й', 2695494),
 ('ф', 424631),
 ('щ', 838272),
 ('ъ', 57419)]

### Удаление из папки всех файлов с "плохой" кодировкой

In [19]:
import chardet
data=r'C:\\Users\\Big data\\New\\all_k'
not_correct = []

bad_files = "bad_file"

if not os.path.exists(os.path.dirname(bad_files)):
    os.makedirs(bad_files)

files = os.listdir(data)

for file in files:
    file_full = os.path.join(data, file)
    with open(file_full, 'rb') as f:
        rawdata = b''.join([f.readline() for _ in range(20)])
    encoding = chardet.detect(rawdata)['encoding']
#    print(f"{file} -> {encoding}")
    if encoding != 'windows-1251':
        not_correct.append((file, encoding))
        os.rename(file_full, os.path.join(os.getcwd(), bad_files, file))

### 2) Посчитать слова в тексте

In [20]:
import string

In [22]:
punc = string.punctuation.replace('-', '')
table = str.maketrans('', '', punc)

In [29]:
files = [os.path.join('C:\\Users\\Big data\\New\\all_k', i) for i in os.listdir('C:\\Users\\Big data\\New\\all_k')]

In [25]:
def count_words(file_name):
    with open(file_name, encoding='ANSI') as f:
        count = Counter(chain.from_iterable(re.findall('[а-я-]+', line.lower()) for line in f))
    return count

In [30]:
%%time
db.from_sequence(files).map(count_words).fold(op.add).compute()

Wall time: 1min 41s


Counter({'хулио': 37,
         'кортасар': 3,
         'модель': 637,
         'для': 100462,
         'сборки': 39,
         'многие': 5689,
         'читатели': 127,
         'наверно': 2197,
         'заметят': 289,
         'что': 715362,
         'в': 1146907,
         'этом': 55443,
         'произведении': 40,
         'я': 568401,
         'кое-где': 483,
         'преступаю': 2,
         'литературные': 50,
         'условности': 96,
         'приведу': 204,
         'лишь': 35228,
         'несколько': 43338,
         'примеров': 99,
         'ар-': 6,
         'гентинцы': 1,
         'у': 203694,
         'меня': 134990,
         'то': 139194,
         'и': 1456575,
         'дело': 27428,
         'переходят': 91,
         'с': 505203,
         'вы': 143989,
         'на': 754898,
         'ты': 211900,
         'когда': 148321,
         'это': 307176,
         'них': 42472,
         'естественно': 2856,
         'диалоге': 15,
         'житель': 222,
         'лондона': 11

In [42]:
%%time
db.from_sequence(files).map(count_words).flatten().frequencies().compute()

Wall time: 1min 52s


[('хулио', 7),
 ('кортасар', 2),
 ('модель', 200),
 ('для', 597),
 ('сборки', 22),
 ('многие', 573),
 ('читатели', 68),
 ('наверно', 208),
 ('заметят', 196),
 ('что', 597),
 ('в', 597),
 ('этом', 597),
 ('произведении', 28),
 ('я', 597),
 ('кое-где', 227),
 ('преступаю', 2),
 ('литературные', 31),
 ('условности', 69),
 ('приведу', 150),
 ('лишь', 595),
 ('несколько', 597),
 ('примеров', 71),
 ('ар-', 2),
 ('гентинцы', 1),
 ('у', 597),
 ('меня', 596),
 ('то', 597),
 ('и', 597),
 ('дело', 597),
 ('переходят', 83),
 ('с', 597),
 ('вы', 596),
 ('на', 597),
 ('ты', 594),
 ('когда', 597),
 ('это', 597),
 ('них', 597),
 ('естественно', 514),
 ('диалоге', 10),
 ('житель', 136),
 ('лондона', 245),
 ('только', 597),
 ('недавно', 568),
 ('бравший', 5),
 ('уроки', 260),
 ('французского', 179),
 ('вдруг', 589),
 ('начинает', 484),
 ('говорить', 595),
 ('нем', 594),
 ('поразительной', 157),
 ('беглостью', 2),
 ('более', 596),
 ('того', 596),
 ('еще', 597),
 ('переводе', 103),
 ('испанский', 70),
 ('

In [33]:
def count_words2(file_name):
    with open(file_name, encoding='ANSI') as f:
        count = Counter(chain.from_iterable(line.translate(table).lower().split() for line in f))
    return count

In [34]:
%%time
db.from_sequence(files).map(count_words2).fold(op.add).compute()

Wall time: 1min 58s


Counter({'хулио': 35,
         'кортасар': 3,
         '62': 51,
         'модель': 634,
         'для': 100344,
         'сборки': 39,
         'многие': 5683,
         'читатели': 126,
         'наверно': 2183,
         'заметят': 288,
         'что': 714234,
         'в': 1145778,
         'этом': 55330,
         'произведении': 40,
         'я': 565111,
         'кое-где': 483,
         'преступаю': 2,
         'литературные': 49,
         'условности': 96,
         'приведу': 204,
         'лишь': 35205,
         'несколько': 43325,
         'примеров': 99,
         'ар-': 5,
         'гентинцы': 1,
         'у': 202996,
         'меня': 134477,
         'то': 138641,
         'и': 1454859,
         'дело': 27289,
         'переходят': 91,
         'с': 503180,
         'вы': 143487,
         'на': 754515,
         'ты': 211046,
         'когда': 148062,
         'это': 306059,
         'них': 42409,
         'естественно': 2845,
         'диалоге': 15,
         'житель': 221,
   

### Для одного из файлов с отзывами из https://yadi.sk/d/OUpjPvLX0i1Ipw подсчитать частоту слов в отзывах с оценкой "5"

In [43]:
filename = os.path.join('Toys_and_Games_5.json')
lines = db.read_text(filename, blocksize=5000000)
lines.take(1)

('{"reviewerID": "A1VXOAVRGKGEAK", "asin": "0439893577", "reviewerName": "Angie", "helpful": [0, 0], "reviewText": "I like the item pricing. My granddaughter wanted to mark on it but I wanted it just for the letters.", "overall": 5.0, "summary": "Magnetic board", "unixReviewTime": 1390953600, "reviewTime": "01 29, 2014"}\n',)

In [44]:
import json
js = lines.map(json.loads)
js.take(1)

({'reviewerID': 'A1VXOAVRGKGEAK',
  'asin': '0439893577',
  'reviewerName': 'Angie',
  'helpful': [0, 0],
  'reviewText': 'I like the item pricing. My granddaughter wanted to mark on it but I wanted it just for the letters.',
  'overall': 5.0,
  'summary': 'Magnetic board',
  'unixReviewTime': 1390953600,
  'reviewTime': '01 29, 2014'},)

In [45]:
perfect = js.filter(lambda x: x['overall'] == 5.0).pluck('reviewText')
perfect.take(2)

('I like the item pricing. My granddaughter wanted to mark on it but I wanted it just for the letters.',
 "Both sides are magnetic.  A real plus when you're entertaining more than one child.  The four-year old can find the letters for the words, while the two-year old can find the pictures the words spell.  (I bought letters and magnetic pictures to go with this board).  Both grandkids liked it a lot, which means I like it a lot as well.  Have not even introduced markers, as this will be used strictly as a magnetic board.")

In [46]:
%%time
result = js.filter(lambda record: record['overall'] == 5.0).\
            map(lambda x: re.sub('[!@#$.?:;()"]',"",x['reviewText'].lower().strip()).split()).\
            flatten().frequencies(sorted == True).compute() #flatten = concat

Wall time: 10.2 s


In [47]:
sorted_by_second = sorted(result, key=lambda tup: tup[1], reverse=True)
sorted_by_second[:20]

[('the', 418907),
 ('and', 292860),
 ('a', 229376),
 ('to', 226557),
 ('it', 190867),
 ('is', 159242),
 ('i', 148458),
 ('this', 145956),
 ('for', 141604),
 ('of', 140828),
 ('with', 108151),
 ('my', 100988),
 ('that', 83632),
 ('in', 82979),
 ('are', 69620),
 ('you', 63555),
 ('on', 61394),
 ('have', 55544),
 ('but', 54069),
 ('as', 53613)]

In [50]:
st1 = """Loved playing Dirt 2 and I thought the graphics were good.  Purchased Dirt 3 as an addition to the other...and the graphics are absolutely "Gorgeous"  If you liked Dirt or Dirt 2...you are going to love Dirt 3.  The game was easier to configure with my Logitech wireless rumblepad...and with my EVGA GTX 580, and all detail set to full on graphics at 1920 x 1080 I get over 100 fps.  The game looks good, plays well and is a blast!"""
st1

'Loved playing Dirt 2 and I thought the graphics were good.  Purchased Dirt 3 as an addition to the other...and the graphics are absolutely "Gorgeous"  If you liked Dirt or Dirt 2...you are going to love Dirt 3.  The game was easier to configure with my Logitech wireless rumblepad...and with my EVGA GTX 580, and all detail set to full on graphics at 1920 x 1080 I get over 100 fps.  The game looks good, plays well and is a blast!'

In [49]:
regex = re.compile('[!@#$.?:;"]')
regex.sub("", st1)

'Loved playing Dirt 2 and I thought the graphics were good  Purchased Dirt 3 as an addition to the otherand the graphics are absolutely Gorgeous  If you liked Dirt or Dirt 2you are going to love Dirt 3  The game was easier to configure with my Logitech wireless rumblepadand with my EVGA GTX 580, and all detail set to full on graphics at 1920 x 1080 I get over 100 fps  The game looks good, plays well and is a blast'

### Для одного из файлов с отзывами из https://yadi.sk/d/OUpjPvLX0i1Ipw подсчитать частоту (!) слов для отзывов с оценкой "5" и для отзывов с оценкой "2" и найти 50 распространенных слов (встречающихся в каждой из групп более 10 раз), частота которых максимально увеличивается и уменьшается при переходе от плохих отзывов к хорошим.

In [119]:
%%time
scores = dict(js.filter(lambda x: x['overall'] == 5).\
                 map(lambda x: re.findall(r'[a-z]+', x['reviewText'].lower())).flatten().frequencies().compute())

Wall time: 10.5 s


In [120]:
for i in list(scores):
    if scores[i] < 10: scores.pop(i)

In [122]:
sum_pos = sum(scores.values())
len(scores)

14270

In [123]:
for i in list(scores.keys()):
    scores[i] /= sum_pos

In [124]:
neg_scores = dict(js.filter(lambda x: x['overall'] == 2).\
                 map(lambda x: re.findall(r'[a-z]+', x['reviewText'].lower())).flatten().frequencies().compute())

In [125]:
for i in list(neg_scores):
    if neg_scores[i] < 10: neg_scores.pop(i)
len(neg_scores)

4186

In [127]:
neg_pos_sum = sum(neg_scores.values())
for i in list(neg_scores.keys()):
    neg_scores[i] /= neg_pos_sum

In [130]:
common_keys = set(neg_scores.keys()).intersection(set(scores.keys()))
common_dict = dict()
for i in common_keys:
    common_dict[i] = scores[i] / neg_scores[i]

In [131]:
b = sorted(list(common_dict.items()), key=lambda x: x[1])
b[-50:]

[('grandson', 3.294867827494073),
 ('beautifully', 3.324381519146118),
 ('tea', 3.333992945863909),
 ('hilarious', 3.3760487983802903),
 ('amazed', 3.3985450778554567),
 ('dragon', 3.3985450778554567),
 ('grandchildren', 3.439953064801504),
 ('class', 3.4418414887501214),
 ('doctor', 3.4526432737362134),
 ('unique', 3.4788889331239066),
 ('duplo', 3.4909405114141747),
 ('daily', 3.4909405114141747),
 ('grandkids', 3.495359423453939),
 ('grand', 3.5213819054658875),
 ('addition', 3.5547692408774436),
 ('love', 3.6156720111412555),
 ('granddaughter', 3.630508129303775),
 ('awesome', 3.6817792401644556),
 ('trips', 3.689791553203589),
 ('beautiful', 3.7071515647883797),
 ('resources', 3.730824307858548),
 ('elf', 3.741345527000846),
 ('math', 3.7503114354873244),
 ('visit', 3.7526760706925497),
 ('favorites', 3.765895038332872),
 ('teaches', 3.773366628738271),
 ('skills', 3.79710798845502),
 ('occupied', 3.840436281831934),
 ('pillow', 3.844453474595357),
 ('expansion', 3.877595314893592

### Для одного из файлов с отзывами из https://yadi.sk/d/OUpjPvLX0i1Ipw найти 5 дней в которые было сделано больше всего отзывов и вернуть количество отзывов в каждый из дней.

In [134]:
%%time
days = js.map(lambda x: x['reviewTime']).frequencies().compute()
top_days = sorted(days, key=lambda x: x[1], reverse=True)[:5]

Wall time: 2.94 s


In [135]:
top_days

[('12 26, 2013', 887),
 ('12 28, 2012', 878),
 ('12 27, 2013', 842),
 ('12 30, 2013', 827),
 ('12 26, 2012', 819)]

### Для одного из файлов с отзывами из https://yadi.sk/d/OUpjPvLX0i1Ipw найти все summary которые были исопльзованы более одного раза. Вернуть эти саммори и количество их повторений в порядке убывания повторений.

In [136]:
%%time
summary = js.map(lambda x: x['summary']).frequencies().compute()

Wall time: 11.6 s


In [137]:
sorted(summary, key=lambda x: x[1],reverse=True)

[('Cute', 558),
 ('Great', 532),
 ('Five Stars', 509),
 ('Fun', 498),
 ('Great toy', 424),
 ('Great!', 388),
 ('Love it', 373),
 ('Fun!', 333),
 ('Love it!', 319),
 ('Great toy!', 309),
 ('Awesome', 287),
 ('cute', 248),
 ('Awesome!', 247),
 ('Perfect', 237),
 ('great', 233),
 ('Cute!', 227),
 ('love it', 197),
 ('Very cute', 194),
 ('Great Toy', 194),
 ('Nice', 192),
 ('Fun toy', 184),
 ('Adorable', 178),
 ('Great game', 177),
 ('fun', 174),
 ('Lots of fun', 163),
 ('Adorable!', 160),
 ('Great product', 155),
 ('Gift', 149),
 ('great toy', 148),
 ('Perfect!', 146),
 ('Great gift', 145),
 ('Great Game', 136),
 ('Good', 132),
 ('So cute!', 125),
 ('So much fun!', 124),
 ('Great fun', 124),
 ('Excellent', 115),
 ('Fun game', 113),
 ('Disappointed', 111),
 ('Great Toy!', 105),
 ('Great Product', 104),
 ('ravage 24/7', 103),
 ('Love', 102),
 ('Great Gift', 102),
 ('Kids love it', 102),
 ('Great game!', 101),
 ('gift', 100),
 ('Cute toy', 99),
 ('Great set', 97),
 ('Lots of fun!', 96),
 ('S

# DataFrame

In [3]:
import os
import dask
import dask.dataframe as dd

In [4]:
import pandas as pd
import numpy as np

In [149]:
from prep import accounts_csvs
# генерация искуственного набора данных:
accounts_csvs(24, 1000000, 500)

Create CSV accounts for dataframe exercise


In [5]:
filename = os.path.join('data', 'accounts.*.csv')
filename

'data\\accounts.*.csv'

In [7]:
# Чтение всех файлов CSV, соответствующих шаблону, в Dask DataFrame:
df = dd.read_csv(filename)
df.head()

id   names  amount
0  466     Bob     261
1  200  Hannah     412
2  338  Yvonne       6
3  351     Bob    2813
4   64  Xavier     313

### 1) В accounts.*.csv найти имя, для которого встречается наибольшее количество положительных значений в столбце amount.

In [8]:
%%time
res = df[df.amount > 0].groupby('names').amount.count().idxmax().compute()

Wall time: 8 s


In [9]:
res

'Ingrid'

### 2) Эффективно создать csv файлы, содержащие данные из файлов `accounts.*.csv` к которым добавлен столбец, в котором содержится количество сотен из положительного значения amount и 0, если значение в amount отрицательное. 

Пример:

`id,names,amount,new
412,Norbert,1492,14
498,Bob,-91,0
448,Ray,72,0`

In [10]:
%%time
df['new'] = df['amount'] // 100
df['new'] = df['new'].where(df['new'] > 0, 0)
df.to_csv('*new.csv')

Wall time: 2min 35s


### 3) Эффективно найти все 10 децилей для распределения значения amount. 

In [11]:
%%time
df['amount'].quantile(np.linspace(0.1, 0.9, 9)).compute()

Wall time: 8.47 s


0.1     -64.0
0.2      20.0
0.3     109.0
0.4     235.0
0.5     447.0
0.6     703.0
0.7    1021.0
0.8    1739.0
0.9    3111.0
Name: amount, dtype: float64

### 4) Найти 2 последовательно идующие транзакции имеющие наибольшую по модулю разность между значениями amount.

In [12]:
%%time
b = df['amount'].map_overlap(lambda data: data.diff(periods=-1), 0, 1).abs().idxmax().compute()
b

251445

### 5) Перевести старый джейсон в датафрейм и посчитать по дням количество

In [13]:
import json

In [14]:
new_df = db.read_text('Toys_and_Games_5.json').map(json.loads)

In [15]:
new_df_1 = new_df.to_dataframe()

In [17]:
%%time
new_df_1.groupby('reviewTime').count()['asin'].nlargest(5).compute()

Wall time: 3.82 s


reviewTime
12 26, 2013    887
12 28, 2012    878
12 27, 2013    842
12 30, 2013    827
12 26, 2012    819
Name: asin, dtype: int64

### 6) Найти количество комментариев по месяцам

In [18]:
new_df_1['month'] = new_df_1['reviewTime'].map(lambda x: x.split()[0]).count()
#тут, конечно, можно по-разному решать. привести к дейттайму и потом группировать, к примеру

In [19]:
new_df_1.groupby('month')['asin'].count().compute()

month
01    27533
02    13925
03    12135
04    10795
05    10530
06    10187
07    10650
08     9150
09     9505
10    11249
11    13942
12    27996
Name: asin, dtype: int64

In [23]:
new_df_1['reviewTime'].map(lambda x: x.split()[0]).value_counts().compute() # другой пример

12    27996
01    27533
11    13942
02    13925
03    12135
10    11249
04    10795
07    10650
05    10530
06    10187
09     9505
08     9150
Name: reviewTime, dtype: int64